### PR TITLE
[ANE-2371] Add Gem constraint

### PR DESCRIPTION
**Pull request by james-fossa**
# Overview

Implement the Gem [Constraint](https://github.com/fossas/locator-rs/pull/14).

## Acceptance criteria

- For gems obeying Semver, this constraint should work just like the fallback does already
- For gems with totally inscrutable versions, this constraint should work just like the fallback does already
- For sequences of arbitrarily long numerics e.g. `1.2.3.4`, this should treat `~>` constraints as binding to the N-1th position. So `1.2.3.4.5 ~> 1.2.3.4`, but `1.2.3.5` does not, for example.
- For prereleases, like `1.2.3prerelease0`, ('prerelease' is not a magic string, any alphabetic string is valid there), this should treat them identically to `1.2.3.prerelease.0`. `1.2.3.prerelease.0` should be less than `1.2.3.prerelease.1` and the numeric portion after the prerelease tag matters.

## Testing plan

I have validated this with unit tests, but have some follow-up questions around some of the cases which I believe to be correct but do not 100% line up with the fallback.

## Metrics

N/A

## Risks

Some of the ruby docs and comments indicate that only "a-z" are supported as valid prerelease tag characters, and I don't know if this is strict enough in that regard to be entirely correct. Namely, I don't know if ruby believe that "a-z" contains, A-Z (I'd think not but it's a comment in a sourcefile in rubygems and the regexes indicate otherwise as I read them.)

Point being, I'm not 100% sure how sure Bundler is of its semantics, so if any edge-cases occur to the reader, please bring them up and I'll add a test.

## References

- [rubygems version class](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/version.rb)
- [rubygems pattern documentation](https://guides.rubygems.org/patterns/)
- [existing scraper code](https://github.com/fossas/sparkle/blob/main/worker/src/scrape/ruby.rs)

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).

### CODE REVIEW

File: src/constraint/gem.rs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+
+    #[test_case(constraint!(Equal => 1,2,3), Revision::from("1*"); "1.2.3_equal_1*")]
+    #[test_case(constraint!(Equal => "1.2.a.0"), Revision::from("1.2*a0"); "1.2.a.0_equal_1.2*a0")]
+    #[test_case(constraint!(Equal => 1,2,3), Revision::from("1*2*3"); "1.2.3_equal_1*2*3")]
+    #[test]
+    fn compare_ruby_error_cases(constraint: Constraint, target: Revision) {
+        assert!(compare(&constraint, FETCHER, &target).is_err())
+    }
+
+    // Testing that we produce the same outputs as our fallback for semvers.
```
Comment (james-fossa) [inline src/constraint/gem.rs:311]: Dunno if these tests are worth their runtime tbh - it's an intentional choice to perform the fallback behavior, so verifying that may be valuable, but readers can read: we call the fallback code so it's sort of tautologically so.

File: src/constraint.rs [hunk 2]
```diff
@@ -109,6 +111,10 @@ impl Constraint {
     ///   In this instance [`Constraint::Compatible`] is a case-insensitive equality comparison.
     pub fn compare(&self, fetcher: Fetcher, target: &Revision) -> bool {
         match fetcher {
+            Fetcher::Gem => gem::compare(self, Fetcher::Gem, target).unwrap_or_else(|err| {
+                warn!("could not parse rubygems version '{err:?}'");
```
Comment (jssblck) [inline src/constraint.rs:115]: [nit] Prefer to include the error and associated metadata as structured output:
```suggestion
                warn!(?err, "could not compare version");
```

Ideally this is combined with a `#[tracing::instrument]` on the `compare` method so that we know what is being compared; I just forgot to add this but feel free to do so.

**Review by jssblck — COMMENTED** [review]
This looks good so far but is missing a critical part: how do we go from the constraints that the gem crawler(s) store in the DB to the `Constraint` types used in this module?

Please add a parser for that as well and I think we're good.

File: src/constraint/gem.rs (hunk 1 above)
Comment (jssblck) [reply src/constraint/gem.rs:311]: Automated tests are mostly about backwards compatibility; I think it makes sense to validate that fallbacks are being used as expected but I'd say you could get away with testing fewer cases that are already effectively tested.

File: src/constraint/gem.rs [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+    match constraints(input.trim()) {
+        Ok((remaining, parsed_constraints)) => {
+            if !remaining.is_empty() {
+                return Err(GemConstraintError::ConstraintsParseError {
+                    constraints: input.to_string(),
+                    message: format!("trailing text: '{remaining}'"),
+                });
+            }
+            Ok(parsed_constraints)
+        }
```
Comment (jssblck) [inline src/constraint/gem.rs:181]: I think you can accomplish the same thing with the `nom` `eof` parser, but this is fine also

**Review by jssblck — APPROVED** [review]
Very nice, great docs too!

File: src/constraint/gem.rs (hunk 3 above)
Comment (james-fossa) [reply src/constraint/gem.rs:181]: Oh my gosh that's exactly what I was looking for - thanks
